File: app/src/main/java/com/china/activity/ForecastActivity.java
package com.china.activity;

/**
 * 天气预报
 */

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.china.R;
import com.china.adapter.WeeklyForecastAdapter;
import com.china.dto.WeatherDto;
import com.china.utils.CommonUtil;
import com.china.utils.OkHttpUtil;
import com.china.utils.WeatherUtil;
import com.china.view.ScrollviewListview;
import com.china.view.WeeklyView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import cn.com.weather.api.WeatherAPI;
import cn.com.weather.beans.Weather;
import cn.com.weather.constants.Constants.Language;
import cn.com.weather.listener.AsyncResponseHandler;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

public class ForecastActivity extends BaseActivity implements OnClickListener{
	
	private Context mContext = null;
	private LinearLayout llBack = null;
	private TextView tvTitle = null;
	private TextView tvCity = null;
	private TextView tvTime = null;//更新时间
	private TextView tvRain = null;//下雨、下雪信息
	private TextView tvTemp = null;//实况温度
	private TextView tvForeTemp = null;//预报温度
	private ImageView ivPhenomenon = null;//天气显现对应的图标
	private TextView tvPhe = null;
	private TextView tvHumidity = null;//相对湿度
	private TextView tvWind = null;//风速
	private TextView tvQuality = null;//空气质量
	private LinearLayout llContainer1 = null;//加载逐小时预报曲线容器
	private ImageView ivSwitcher = null;//列表和趋势开关
	private ScrollviewListview mListView = null;//一周预报列表listview
	private WeeklyForecastAdapter mAdapter = null;
	private LinearLayout llContainer2 = null;//一周预报曲线容器
	private ScrollView scrollView = null;//全屏
	private int width = 0;
	private List<WeatherDto> weeklyList = new ArrayList<>();
	private SimpleDateFormat sdf1 = new SimpleDateFormat("yyyyMMddHHmm");
	private SimpleDateFormat sdf2 = new SimpleDateFormat("HH");
	private SimpleDateFormat sdf3 = new SimpleDateFormat("yyyyMMdd");
	private ImageView ivShare = null;
	private RelativeLayout reLocation = null;
	private LinearLayout llThree = null;
	private RelativeLayout reWeekly = null;
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_forecast);
		mContext = this;
		showDialog();
		initWidget();
		initListView();
	}
	
	/**
	 * 初始化控件
	 */
	private void initWidget() {
		llBack = (LinearLayout) findViewById(R.id.llBack);
		llBack.setOnClickListener(this);
		tvTitle = (TextView) findViewById(R.id.tvTitle);
		//解决scrollView嵌套listview，动态计算listview高度后，自动滚动到屏幕底部
		tvCity = (TextView) findViewById(R.id.tvCity);
		tvTime = (TextView) findViewById(R.id.tvTime);
		tvTime.setFocusable(true);
		tvTime.setFocusableInTouchMode(true);
		tvTime.requestFocus();
		tvRain = (TextView) findViewById(R.id.tvRain);
		ivSwitcher = (ImageView) findViewById(R.id.ivSwitcher);
		ivSwitcher.setOnClickListener(this);
		tvTemp = (TextView) findViewById(R.id.tvTemp);
		tvForeTemp = (TextView) findViewById(R.id.tvForeTemp);
		ivPhenomenon = (ImageView) findViewById(R.id.ivPhenomenon);
		tvPhe = (TextView) findViewById(R.id.tvPhe);
		tvQuality = (TextView) findViewById(R.id.tvQuality);
		tvHumidity = (TextView) findViewById(R.id.tvHumidity);
		tvWind = (TextView) findViewById(R.id.tvWind);
		llContainer1 = (LinearLayout) findViewById(R.id.llContainer1);
		llContainer2 = (LinearLayout) findViewById(R.id.llContainer2);
		scrollView = (ScrollView) findViewById(R.id.scrollView);
		ivShare = (ImageView) findViewById(R.id.ivShare);
		ivShare.setOnClickListener(this);
		ivShare.setVisibility(View.VISIBLE);
		reLocation = (RelativeLayout) findViewById(R.id.reLocation);
		llThree = (LinearLayout) findViewById(R.id.llThree);
		reWeekly = (RelativeLayout) findViewById(R.id.reWeekly);
		
		DisplayMetrics dm = new DisplayMetrics();
		getWindowManager().getDefaultDisplay().getMetrics(dm);
		width = dm.widthPixels;
		
		refresh();
	}
	
	private void refresh() {
		String areaName = getIntent().getStringExtra("cityName");
		if (!TextUtils.isEmpty(areaName)) {
			tvTitle.setText(areaName);
			tvCity.setText(areaName);
		}
		
		String cityId = getIntent().getStringExtra("cityId");
		if (!TextUtils.isEmpty(cityId)) {
			getWeatherInfo(cityId);
		}
	}

	/**
	 * 获取天气数据
	 */
	private void getWeatherInfo(final String cityId) {
		new Thread(new Runnable() {
			@Override
			public void run() {
				WeatherAPI.getWeather2(mContext, cityId, Language.ZH_CN, new AsyncResponseHandler() {
					@Override
					public void onComplete(final Weather content) {
						super.onComplete(content);
						runOnUiThread(new Runnable() {
							@Override
							public void run() {
								if (content != null) {
									try {
										//实况信息
										JSONObject object = content.getWeatherFactInfo();
										if (!object.isNull("l7")) {
											String time = object.getString("l7");
											if (time != null) {
												tvTime.setText(time + getString(R.string.publish));
											}
										}
										if (!object.isNull("l5")) {
											String weatherCode = WeatherUtil.lastValue(object.getString("l5"));
											int current = Integer.parseInt(sdf2.format(new Date()));
											if (current >= 5 && current < 18) {
												ivPhenomenon.setImageBitmap(WeatherUtil.getDayBitmap(mContext, Integer.valueOf(weatherCode)));
											}else {
												ivPhenomenon.setImageBitmap(WeatherUtil.getNightBitmap(mContext, Integer.valueOf(weatherCode)));
											}
											tvPhe.setText(getString(WeatherUtil.getWeatherId(Integer.valueOf(weatherCode))));
										}
										if (!object.isNull("l1")) {
											String factTemp = WeatherUtil.lastValue(object.getString("l1"));
											tvTemp.setText(factTemp);
										}
										if (!object.isNull("l2")) {
											String humidity = WeatherUtil.lastValue(object.getString("l2"));
											tvHumidity.setText(getString(R.string.humidity)+" "+humidity + getString(R.string.unit_percent));
										}
										if (!object.isNull("l4")) {
											String windDir = WeatherUtil.lastValue(object.getString("l4"));
											if (!object.isNull("l3")) {
												String windForce = WeatherUtil.lastValue(object.getString("l3"));
												if (!TextUtils.isEmpty(windDir) && !TextUtils.isEmpty(windForce)) {
													tvWind.setText(getString(WeatherUtil.getWindDirection(Integer.valueOf(windDir)))+
															" " + WeatherUtil.getFactWindForce(Integer.valueOf(windForce)));
												}
											}
										}

										//空气质量
										JSONObject obj = content.getAirQualityInfo();
										if (obj != null) {
											if (!obj.isNull("k3")) {
												String num = WeatherUtil.lastValue(obj.getString("k3"));
												if (!TextUtils.isEmpty(num)) {
													tvQuality.setText(getString(R.string.weather_quality) + " " +
															WeatherUtil.getAqi(mContext, Integer.valueOf(num)) + " " + num);
												}
											}
										}

										//逐小时预报信息
										JSONArray hourlyArray = content.getHourlyFineForecast2();
										llContainer1.removeAllViews();
										for (int i = 0; i < hourlyArray.length(); i++) {
											JSONObject itemObj = hourlyArray.getJSONObject(i);
											int hourlyCode = Integer.valueOf(itemObj.getString("ja"));
											int hourlyTemp = Integer.valueOf(itemObj.getString("jb"));
											String hourlyTime = itemObj.getString("jf");

											LayoutInflater inflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
											View view = inflater.inflate(R.layout.layout_hour_forecast, null);
											TextView tvHour = (TextView) view.findViewById(R.id.tvHour);
											ImageView ivPhe = (ImageView) view.findViewById(R.id.ivPhe);
											TextView tvPhe = (TextView) view.findViewById(R.id.tvPhe);
											TextView tvTemp = (TextView) view.findViewById(R.id.tvTemp);
											try {
												int current = Integer.parseInt(sdf2.format(sdf1.parse(hourlyTime)));
												tvHour.setText(current+getString(R.string.hour));
												if (current >= 5 && current < 18) {
													ivPhe.setImageBitmap(WeatherUtil.getDayBitmap(mContext, hourlyCode));
												}else {
													ivPhe.setImageBitmap(WeatherUtil.getNightBitmap(mContext, hourlyCode));
												}
											} catch (ParseException e) {
												e.printStackTrace();
											}
											tvPhe.setText(getString(WeatherUtil.getWeatherId(hourlyCode)));
											tvTemp.setText(hourlyTemp+getString(R.string.unit_degree));
											llContainer1.addView(view);
										}

										//一周预报信息
										long foreDate = 0;
										long currentDate = 0;
										try {
											String f0 = sdf3.format(sdf1.parse(content.getForecastTime()));
											foreDate = sdf3.parse(f0).getTime();
											currentDate = sdf3.parse(sdf3.format(new Date())).getTime();
										} catch (ParseException e) {
											e.printStackTrace();
										}

										weeklyList.clear();
										//这里只去一周预报，默认为15天，所以遍历7次
										for (int i = 1; i <= 15; i++) {
											WeatherDto dto = new WeatherDto();

											JSONArray weeklyArray = content.getWeatherForecastInfo(i);
											JSONObject weeklyObj = weeklyArray.getJSONObject(0);

											//晚上
											dto.lowPheCode = Integer.valueOf(weeklyObj.getString("fb"));
											dto.lowPhe = getString(WeatherUtil.getWeatherId(Integer.valueOf(weeklyObj.getString("fb"))));
											dto.lowTemp = Integer.valueOf(weeklyObj.getString("fd"));
											dto.lowWindDir = Integer.valueOf(weeklyObj.getString("ff"));
											dto.lowWindForce = Integer.valueOf(weeklyObj.getString("fh"));

											//白天数据缺失时，就使用第二天白天数据
											if (TextUtils.isEmpty(weeklyObj.getString("fa"))) {
												JSONObject secondObj = content.getWeatherForecastInfo(2).getJSONObject(0);
												dto.highPheCode = Integer.valueOf(secondObj.getString("fa"));
												dto.highPhe = getString(WeatherUtil.getWeatherId(Integer.valueOf(secondObj.getString("fa"))));

												int time1 = Integer.valueOf(secondObj.getString("fc"));
												int time2 = Integer.valueOf(weeklyObj.getString("fd"));
												if (time1 <= time2) {
													dto.highTemp = time2 + 2;
												}else {
													dto.highTemp = Integer.valueOf(secondObj.getString("fc"));
												}

												dto.highWindDir = Integer.valueOf(secondObj.getString("fe"));
												dto.highWindForce = Integer.valueOf(secondObj.getString("fg"));
											}else {
												//白天
												dto.highPheCode = Integer.valueOf(weeklyObj.getString("fa"));
												dto.highPhe = getString(WeatherUtil.getWeatherId(Integer.valueOf(weeklyObj.getString("fa"))));
												dto.highTemp = Integer.valueOf(weeklyObj.getString("fc"));
												dto.highWindDir = Integer.valueOf(weeklyObj.getString("fe"));
												dto.highWindForce = Integer.valueOf(weeklyObj.getString("fg"));
											}

											JSONArray timeArray =  content.getTimeInfo(i);
											JSONObject timeObj = timeArray.getJSONObject(0);
											dto.week = timeObj.getString("t4");//星期几
											dto.date = timeObj.getString("t1");//日期

											weeklyList.add(dto);

											if (currentDate > foreDate) {
												if (i == 2) {
													tvForeTemp.setText(dto.highTemp+"℃"+"/"+dto.lowTemp+"℃");
												}
											}else {
												if (i == 1) {
													tvForeTemp.setText(dto.highTemp+"℃"+"/"+dto.lowTemp+"℃");
												}
											}
										}

										//一周预报列表
										if (mAdapter != null) {
											mAdapter.foreDate = foreDate;
											mAdapter.currentDate = currentDate;
											mAdapter.notifyDataSetChanged();
										}

										//一周预报曲线
										WeeklyView weeklyView = new WeeklyView(mContext);
										weeklyView.setData(weeklyList, foreDate, currentDate);
										llContainer2.removeAllViews();
										llContainer2.addView(weeklyView, width*2, (int)(CommonUtil.dip2px(mContext, 400)));
									} catch (JSONException e) {
										e.printStackTrace();
									} catch (NullPointerException e) {
										e.printStackTrace();
									}

									scrollView.setVisibility(View.VISIBLE);
									cancelDialog();
								}
							}
						});
					}

					@Override
					public void onError(Throwable error, String content) {
						super.onError(error, content);
						cancelDialog();
					}
				});
			}
		}).start();
	}
	
	/**
	 * 初始化listview
	 */
	private void initListView() {
		mListView = (ScrollviewListview) findViewById(R.id.listView);
		mAdapter = new WeeklyForecastAdapter(mContext, weeklyList);
		mListView.setAdapter(mAdapter);
	}
	
	/**
	 * 异步加载一小时内降雨、或降雪信息
	 * @param lng
	 * @param lat
	 */
	private void OkHttpMinuteFall(double lng, double lat) {
		final String url = "http://api.caiyunapp.com/v2/HyTVV5YAkoxlQ3Zd/"+lng+","+lat+"/forecast";
		new Thread(new Runnable() {
			@Override
			public void run() {
				OkHttpUtil.enqueue(new Request.Builder().url(url).build(), new Callback() {
					@Override
					public void onFailure(Call call, IOException e) {

					}

					@Override
					public void onResponse(Call call, Response response) throws IOException {
						if (!response.isSuccessful()) {
							return;
						}
						final String result = response.body().string();
						runOnUiThread(new Runnable() {
							@Override
							public void run() {
								if (!TextUtils.isEmpty(result)) {
									try {
										JSONObject object = new JSONObject(result);
										if (object != null) {
											if (!object.isNull("result")) {
												JSONObject objResult = object.getJSONObject("result");
												if (!objResult.isNull("minutely")) {
													JSONObject objMin = objResult.getJSONObject("minutely");
													if (!objMin.isNull("description")) {
														String rain = objMin.getString("description");
														if (!TextUtils.isEmpty(rain)) {
															tvRain.setText(rain.replace(getString(R.string.little_caiyun), ""));
															tvRain.setVisibility(View.VISIBLE);
														}else {
															tvRain.setVisibility(View.GONE);
														}
													}
												}
											}
										}
									} catch (JSONException e1) {
										e1.printStackTrace();
									}
								}
							}
						});
					}
				});
			}
		}).start();
	}
	
	@Override
	public void onClick(View v) {
		switch (v.getId()) {
		case R.id.llBack:
			finish();
			break;
		case R.id.ivSwitcher:
			if (mListView.getVisibility() == View.VISIBLE) {
				ivSwitcher.setImageResource(R.drawable.iv_trend);
				mListView.setVisibility(View.GONE);
				llContainer2.setVisibility(View.VISIBLE);
			}else {
				ivSwitcher.setImageResource(R.drawable.iv_list);
				mListView.setVisibility(View.VISIBLE);
				llContainer2.setVisibility(View.GONE);
			}
			break;
		case R.id.ivShare:
			Bitmap bitmap;
			if (mListView.getVisibility() == View.VISIBLE) {
				Bitmap bitmap1 = CommonUtil.captureScrollView(scrollView);
				Bitmap bitmap2 = BitmapFactory.decodeResource(getResources(), R.drawable.iv_share_bottom);
				bitmap = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap1, bitmap2, false);
				CommonUtil.clearBitmap(bitmap1);
				CommonUtil.clearBitmap(bitmap2);
			}else {
				Bitmap bitmap1 = CommonUtil.captureView(reLocation);
				Bitmap bitmap2 = CommonUtil.captureView(llThree);
				Bitmap bitmap3 = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap1, bitmap2, false);
				CommonUtil.clearBitmap(bitmap1);
				CommonUtil.clearBitmap(bitmap2);
				Bitmap bitmap4 = CommonUtil.captureView(llContainer1);
				Bitmap bitmap5 = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap3, bitmap4, false);
				CommonUtil.clearBitmap(bitmap3);
				CommonUtil.clearBitmap(bitmap4);
				Bitmap bitmap6 = CommonUtil.captureView(reWeekly);
				Bitmap bitmap7 = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap5, bitmap6, false);
				CommonUtil.clearBitmap(bitmap5);
				CommonUtil.clearBitmap(bitmap6);
				Bitmap bitmap8 = CommonUtil.captureView(llContainer2);
				Bitmap bitmap9 = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap7, bitmap8, false);
				CommonUtil.clearBitmap(bitmap7);
				CommonUtil.clearBitmap(bitmap8);
				Bitmap bitmap10 = BitmapFactory.decodeResource(getResources(), R.drawable.iv_share_bottom);
				bitmap = CommonUtil.mergeBitmap(ForecastActivity.this, bitmap9, bitmap10, false);
				CommonUtil.clearBitmap(bitmap9);
				CommonUtil.clearBitmap(bitmap10);
			}
			CommonUtil.share(ForecastActivity.this, bitmap);
			break;

		default:
			break;
		}
	}

}

File: app/src/main/java/com/china/activity/ForecastActivity.kt
package com.china.activity;

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.widget.ImageView
import android.widget.TextView
import com.china.R
import com.china.adapter.ShawnWeeklyForecastAdapter
import com.china.common.MyApplication
import com.china.dto.WeatherDto
import com.china.utils.AuthorityUtil
import com.china.utils.CommonUtil
import com.china.utils.OkHttpUtil
import com.china.utils.WeatherUtil
import com.china.view.WeeklyView
import kotlinx.android.synthetic.main.activity_forecast.*
import kotlinx.android.synthetic.main.shawn_layout_title.*
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

/**
 * 天气预报
 */
class ForecastActivity : ShawnBaseActivity(), OnClickListener {

	private var mAdapter: ShawnWeeklyForecastAdapter? = null
	private val weeklyList: ArrayList<WeatherDto> = ArrayList()
	private val sdf1 = SimpleDateFormat("yyyyMMddHHmm", Locale.CHINA)
	private val sdf2 = SimpleDateFormat("HH", Locale.CHINA)
	private val sdf3 = SimpleDateFormat("yyyyMMdd", Locale.CHINA)

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_forecast)
		checkAuthority()
	}

	private fun init() {
		initWidget()
		initListView()
	}

	/**
	 * 初始化控件
	 */
	private fun initWidget() {
		showDialog()
		llBack.setOnClickListener(this)
		tvTitle.text = "天气详情"
		//解决scrollView嵌套listview，动态计算listview高度后，自动滚动到屏幕底部
		tvTime.isFocusable = true
		tvTime.isFocusableInTouchMode = true
		tvTime.requestFocus()
		ivSwitcher.setOnClickListener(this)
		ivShare.setOnClickListener(this)
		ivShare.visibility = View.VISIBLE
		refresh()
	}

	private fun refresh() {
		val areaName = intent.getStringExtra("cityName")
		if (!TextUtils.isEmpty(areaName)) {
			tvCity!!.text = areaName
		}
		val cityId = intent.getStringExtra("cityId")
		if (!TextUtils.isEmpty(cityId)) {
			getWeatherInfo(cityId)
		}
	}

	/**
	 * 获取天气数据
	 */
	private fun getWeatherInfo(cityId: String) {
		Thread(Runnable {
			val url = "https://videoshfcx.tianqi.cn/dav_tqwy/ty_weather/data/$cityId.html"
			OkHttpUtil.enqueue(Request.Builder().url(String.format(url, cityId)).build(), object : Callback {
				override fun onFailure(call: Call, e: IOException) {}
				@Throws(IOException::class)
				override fun onResponse(call: Call, response: Response) {
					if (!response.isSuccessful) {
						return
					}
					val result = response.body!!.string()
					runOnUiThread {
						if (!TextUtils.isEmpty(result)) {
							try {
								val obj = JSONObject(result)
								//逐小时预报信息
								if (!obj.isNull("jh")) {
									val jh = obj.getJSONArray("jh")
									llContainer1!!.removeAllViews()
									for (i in 0 until jh.length()) {
										val itemObj = jh.getJSONObject(i)
										val hourlyCode = Integer.valueOf(itemObj.getString("ja"))
										val hourlyTemp = Integer.valueOf(itemObj.getString("jb"))
										val hourlyTime = itemObj.getString("jf")
										val inflater = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
										val view = inflater.inflate(R.layout.shawn_layout_hour_forecast, null)
										val tvHour = view.findViewById<TextView>(R.id.tvHour)
										val ivPhe = view.findViewById<ImageView>(R.id.ivPhe)
										val tvPhe = view.findViewById<TextView>(R.id.tvPhe)
										val tvTemp = view.findViewById<TextView>(R.id.tvTemp)
										try {
											val current = sdf2.format(sdf1.parse(hourlyTime)).toInt()
											tvHour.text = current.toString() + "时"
											if (current in 5..17) {
												ivPhe.setImageBitmap(WeatherUtil.getDayBitmap(this@ForecastActivity, hourlyCode))
											} else {
												ivPhe.setImageBitmap(WeatherUtil.getNightBitmap(this@ForecastActivity, hourlyCode))
											}
										} catch (e: ParseException) {
											e.printStackTrace()
										}
										tvPhe.text = getString(WeatherUtil.getWeatherId(hourlyCode))
										tvTemp.text = hourlyTemp.toString() + getString(R.string.unit_degree)
										llContainer1!!.addView(view)
									}
								}
								//15天预报
								if (!obj.isNull("f")) {
									weeklyList.clear()
									val f = obj.getJSONObject("f")
									val f0 = f.getString("f0")
									var foreDate: Long = 0
									var currentDate: Long = 0
									try {
										val fTime = sdf3.format(sdf1.parse(f0))
										foreDate = sdf3.parse(fTime).time
										currentDate = sdf3.parse(sdf3.format(Date())).time
									} catch (e: ParseException) {
										e.printStackTrace()
									}
									if (!f.isNull("f1")) {
										val f1 = f.getJSONArray("f1")
										for (i in 0 until f1.length()) {
											val dto = WeatherDto()
											val weeklyObj = f1.getJSONObject(i)
											//晚上
											dto.lowPheCode = Integer.valueOf(weeklyObj.getString("fb"))
											dto.lowPhe = getString(WeatherUtil.getWeatherId(Integer.valueOf(weeklyObj.getString("fb"))))
											dto.lowTemp = Integer.valueOf(weeklyObj.getString("fd"))
											dto.lowWindDir = Integer.valueOf(weeklyObj.getString("ff"))
											dto.lowWindForce = Integer.valueOf(weeklyObj.getString("fh"))
											//白天
											dto.highPheCode = Integer.valueOf(weeklyObj.getString("fa"))
											dto.highPhe = getString(WeatherUtil.getWeatherId(Integer.valueOf(weeklyObj.getString("fa"))))
											dto.highTemp = Integer.valueOf(weeklyObj.getString("fc"))
											dto.highWindDir = Integer.valueOf(weeklyObj.getString("fe"))
											dto.highWindForce = Integer.valueOf(weeklyObj.getString("fg"))
											if (currentDate > foreDate) {
												when (i) {
													0 -> {
														dto.week = "昨天"
													}
													1 -> {
														dto.week = "今天"
														tvFactInfo.text = dto.highTemp.toString() + "℃" + "/" + dto.lowTemp + "℃" + "\n"
													}
													2 -> {
														dto.week = "明天"
													}
													else -> {
														dto.week = CommonUtil.getWeek(i - 1) //星期几
													}
												}
												dto.date = CommonUtil.getDate(f0, i) //日期
											} else {
												when (i) {
													0 -> {
														dto.week = "今天"
														tvFactInfo.text = dto.highTemp.toString() + "℃" + "/" + dto.lowTemp + "℃" + "\n"
													}
													1 -> {
														dto.week = "明天"
													}
													else -> {
														dto.week = CommonUtil.getWeek(i) //星期几
													}
												}
												dto.date = CommonUtil.getDate(f0, i) //日期
											}
											weeklyList.add(dto)
										}
									}
									//一周预报列表
									if (mAdapter != null) {
										mAdapter!!.notifyDataSetChanged()
									}
									//一周预报曲线
									val weeklyView = WeeklyView(this@ForecastActivity)
									weeklyView.setData(weeklyList, foreDate, currentDate)
									llContainer2.removeAllViews()
									llContainer2.addView(weeklyView, CommonUtil.widthPixels(this@ForecastActivity) * 2, CommonUtil.dip2px(this@ForecastActivity, 400f).toInt())
									//空气质量
									if (!obj.isNull("k")) {
										val k = obj.getJSONObject("k")
										if (!k.isNull("k3")) {
											val num = WeatherUtil.lastValue(k.getString("k3"))
											if (!TextUtils.isEmpty(num)) {
												tvFactInfo.text = tvFactInfo.text.toString() + "空气质量 " + WeatherUtil.getAqi(this@ForecastActivity, Integer.valueOf(num)) + " " + num + "\n"
											}
										}
									}
									//实况信息
									if (!obj.isNull("l")) {
										val l = obj.getJSONObject("l")
										if (!l.isNull("l7")) {
											val time = l.getString("l7")
											if (time != null) {
												tvTime.text = time + "发布"
											}
										}
										if (!l.isNull("l5")) {
											val weatherCode = WeatherUtil.lastValue(l.getString("l5"))
											val current = sdf2.format(Date()).toInt()
											if (current in 5..17) {
												ivPhenomenon!!.setImageBitmap(WeatherUtil.getDayBitmap(this@ForecastActivity, Integer.valueOf(weatherCode)))
											} else {
												ivPhenomenon!!.setImageBitmap(WeatherUtil.getNightBitmap(this@ForecastActivity, Integer.valueOf(weatherCode)))
											}
											tvPhe.text = getString(WeatherUtil.getWeatherId(Integer.valueOf(weatherCode)))
										}
										if (!MyApplication.FACTENABLE) {
											if (!l.isNull("l1")) {
												tvTemp.text = WeatherUtil.lastValue(l.getString("l1"))
											}
											if (!l.isNull("l2")) {
												val humidity = WeatherUtil.lastValue(l.getString("l2"))
												tvFactInfo.text = tvFactInfo.text.toString() + "相对湿度 " + humidity + getString(R.string.unit_percent) + "\n"
											}
											if (!l.isNull("l4")) {
												val windDir = WeatherUtil.lastValue(l.getString("l4"))
												if (!l.isNull("l3")) {
													val windForce = WeatherUtil.lastValue(l.getString("l3"))
													if (!TextUtils.isEmpty(windDir) && !TextUtils.isEmpty(windForce)) {
														tvFactInfo.text = tvFactInfo.text.toString() + getString(WeatherUtil.getWindDirection(Integer.valueOf(windDir))) +
																" " + WeatherUtil.getFactWindForce(Integer.valueOf(windForce)) + "\n"
													}
												}
											}
										} else {
											okHttpFact()
										}
									}
								}
							} catch (e: JSONException) {
								e.printStackTrace()
							}
						}
						scrollView!!.visibility = View.VISIBLE
						cancelDialog()
					}
				}
			})
		}).start()
	}

	private fun okHttpFact() {
		Thread(Runnable {
			val lat = intent.getDoubleExtra("lat", 0.0)
			val lng = intent.getDoubleExtra("lng", 0.0)
			val url = "https://music.data.cma.cn/lsb/api?elements=TEM,PRE,RHU,WINS,WIND,WEA&interfaceId=getSurfEleInLocationByTime&lat=$lat&lon=$lng&apikey=AxEkluey201exDxyBoxUeYSw&nsukey=IGzynTgkKQ1Hfa3iJTwv4lci%2F%2F13c%2FQm3p83hih8xiri%2Bc5bm0ia85VASrEHrZRsgj6nlBF1U6F3m5PDkUd6oPtd7itR8p%2BwpJi7yIE%2FVcBsCwya6rhj%2BP%2BhBPCCyrb%2BsyYZLhRk5pkL73jJKE%2Ff4O7PWGPRwVtgQAqgFQ1XEXROJp7qMek79o6%2BiukbiCuY"
			OkHttpUtil.enqueue(Request.Builder().url(url).build(), object : Callback {
				override fun onFailure(call: Call, e: IOException) {}
				@Throws(IOException::class)
				override fun onResponse(call: Call, response: Response) {
					if (!response.isSuccessful) {
						return
					}
					val result = response.body!!.string()
					runOnUiThread {
						if (!TextUtils.isEmpty(result)) {
							try {
								val obj = JSONObject(result)
								if (!obj.isNull("DS")) {
									val array = obj.getJSONArray("DS")
									if (array.length() > 0) {
										val itemObj = array.getJSONObject(0)
										if (!itemObj.isNull("TEM")) {
											tvTemp.text = itemObj.getString("TEM")
										}
										if (!itemObj.isNull("RHU")) {
											val humidity = itemObj.getString("RHU")
											tvFactInfo.text = tvFactInfo.text.toString() + "相对湿度 " + humidity + getString(R.string.unit_percent) + "\n"
										}
										if (!itemObj.isNull("WIND")) {
											val windDir = itemObj.getString("WIND").toFloat().toInt()
											val windForce = itemObj.getString("WINS").toFloat().toInt()
											var dir = getString(WeatherUtil.getWindDirection(windDir))
											if (TextUtils.isEmpty(dir)) {
												dir = "无持续风向"
											}
											tvFactInfo.text = tvFactInfo.text.toString() + dir +
													" " + WeatherUtil.getFactWindForce(windForce) + "\n"
										}
									}
								}
							} catch (e: JSONException) {
								e.printStackTrace()
							}
						}
					}
				}
			})
		}).start()
	}

	/**
	 * 初始化listview
	 */
	private fun initListView() {
		mAdapter = ShawnWeeklyForecastAdapter(this, weeklyList)
		listView.adapter = mAdapter
	}

	override fun onClick(v: View) {
		when (v.id) {
			R.id.llBack -> finish()
			R.id.ivSwitcher -> {
				if (listView!!.visibility == View.VISIBLE) {
					ivSwitcher.setImageResource(R.drawable.shawn_icon_switch_trend)
					listView!!.visibility = View.GONE
					llContainer2.visibility = View.VISIBLE
				} else {
					ivSwitcher.setImageResource(R.drawable.shawn_icon_switch_list)
					listView!!.visibility = View.VISIBLE
					llContainer2.visibility = View.GONE
				}
			}
			R.id.ivShare -> {
				val bitmap: Bitmap
				if (listView!!.visibility == View.VISIBLE) {
					val bitmap1 = CommonUtil.captureScrollView(scrollView)
					val bitmap2 = BitmapFactory.decodeResource(resources, R.drawable.shawn_legend_share_portrait)
					bitmap = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap1, bitmap2, false)
					CommonUtil.clearBitmap(bitmap1)
					CommonUtil.clearBitmap(bitmap2)
				} else {
					val bitmap1 = CommonUtil.captureView(clLocation)
					val bitmap2 = CommonUtil.captureView(llHourly)
					val bitmap3 = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap1, bitmap2, false)
					CommonUtil.clearBitmap(bitmap1)
					CommonUtil.clearBitmap(bitmap2)
					val bitmap4 = CommonUtil.captureView(llContainer1)
					val bitmap5 = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap3, bitmap4, false)
					CommonUtil.clearBitmap(bitmap3)
					CommonUtil.clearBitmap(bitmap4)
					val bitmap6 = CommonUtil.captureView(clWeekly)
					val bitmap7 = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap5, bitmap6, false)
					CommonUtil.clearBitmap(bitmap5)
					CommonUtil.clearBitmap(bitmap6)
					val bitmap8 = CommonUtil.captureView(llContainer2)
					val bitmap9 = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap7, bitmap8, false)
					CommonUtil.clearBitmap(bitmap7)
					CommonUtil.clearBitmap(bitmap8)
					val bitmap10 = BitmapFactory.decodeResource(resources, R.drawable.shawn_legend_share_portrait)
					bitmap = CommonUtil.mergeBitmap(this@ForecastActivity, bitmap9, bitmap10, false)
					CommonUtil.clearBitmap(bitmap9)
					CommonUtil.clearBitmap(bitmap10)
				}
				CommonUtil.share(this@ForecastActivity, bitmap)
			}
		}
	}

	//需要申请的所有权限
	var allPermissions = arrayOf(
			Manifest.permission.CALL_PHONE,
			Manifest.permission.WRITE_EXTERNAL_STORAGE
	)

	//拒绝的权限集合
	var deniedList: MutableList<String> = ArrayList()

	/**
	 * 申请定位权限
	 */
	private fun checkAuthority() {
		if (Build.VERSION.SDK_INT < 23) {
			init()
		} else {
			deniedList.clear()
			for (permission in allPermissions) {
				if (ContextCompat.checkSelfPermission(this, permission) !== PackageManager.PERMISSION_GRANTED) {
					deniedList.add(permission)
				}
			}
			if (deniedList.isEmpty()) { //所有权限都授予
				init()
			} else {
				val permissions = deniedList.toTypedArray() //将list转成数组
				ActivityCompat.requestPermissions(this, permissions, AuthorityUtil.AUTHOR_LOCATION)
			}
		}
	}

	override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String?>, grantResults: IntArray) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults)
		when (requestCode) {
			AuthorityUtil.AUTHOR_LOCATION -> if (grantResults.isNotEmpty()) {
				var isAllGranted = true //是否全部授权
				for (gResult in grantResults) {
					if (gResult != PackageManager.PERMISSION_GRANTED) {
						isAllGranted = false
						break
					}
				}
				if (isAllGranted) { //所有权限都授予
					init()
				} else { //只要有一个没有授权，就提示进入设置界面设置
					AuthorityUtil.intentAuthorSetting(this, "\"" + getString(R.string.app_name) + "\"" + "需要使用电话权限、存储权限，是否前往设置？")
				}
			} else {
				for (permission in permissions) {
					if (!ActivityCompat.shouldShowRequestPermissionRationale(this, permission!!)) {
						AuthorityUtil.intentAuthorSetting(this, "\"" + getString(R.string.app_name) + "\"" + "需要使用电话权限、存储权限，是否前往设置？")
						break
					}
				}
			}
		}
	}

}